parser_library/src/processing/opencode_provider.cpp

```diff
@@ -0,0 +1,563 @@
+        result = std::move(m_preprocessor_buffer.back());
+        m_preprocessor_buffer.pop_back();
+    }
+    else if (!m_next_line_text.empty())
+    {
+        result = lexing::extract_line(m_next_line_text).first;
+
+        generate_aread_highlighting(result, m_current_line);
+
+        m_lines_to_remove.current_text_lines++;
+    }
+    else
+        adjust_length = false;
+
+    if (adjust_length)
+        result.resize(80, ' ');
+    return result;
+}
+
+void opencode_provider::ainsert(const std::string& rec, ainsert_destination dest)
+{
+    switch (dest)
+    {
+        case hlasm_plugin::parser_library::processing::ainsert_destination::back:
+            m_ainsert_buffer.push_back(rec);
+            break;
+        case hlasm_plugin::parser_library::processing::ainsert_destination::front:
+            m_ainsert_buffer.push_front(rec);
+            break;
+    }
+}
+
+extract_next_logical_line_result opencode_provider::feed_line(parsing::parser_holder& p)
+{
+    auto ll_res = extract_next_logical_line();
+    if (ll_res == extract_next_logical_line_result::failed)
+        return ll_res;
+    m_line_fed = true;
+
+    p.input->reset(m_current_logical_line);
+
+    p.lex->set_file_offset({ m_current_logical_line_source.begin_line, 0 /*lexing::default_ictl.begin-1 really*/ },
+        ll_res == extract_next_logical_line_result::process);
+    p.lex->set_unlimited_line(false);
+    p.lex->reset();
+
+    p.stream->reset();
+
+    p.parser->reset();
+
+    p.parser->get_collector().prepare_for_next_statement();
+
+    m_ctx->hlasm_ctx->metrics.lines += m_current_logical_line.segments.size();
+
+    return ll_res;
+}
+
+bool opencode_provider::is_comment()
+{
+    auto prefix = m_current_logical_line.segments.front().code.substr(0, 2);
+    return prefix == ".*" || prefix.substr(0, 1) == "*";
+}
+
+void opencode_provider::process_comment()
+{
+    size_t line_no = m_current_logical_line_source.begin_line;
+    for (const auto& l : m_current_logical_line.segments)
+    {
+        if (l.code.size())
+        {
+            auto [skip_len, _] = lexing::length_utf16(l.line.substr(0, l.code.data() - l.line.data()));
+            auto [code_len, last_big] = lexing::length_utf16(l.code);
+
+            m_src_proc->add_hl_symbol(
+                token_info(range(position(line_no, skip_len), position(line_no, skip_len + code_len)),
+                    semantics::hl_scopes::comment));
+        }
+        ++line_no;
+    }
+}
+
+void opencode_provider::generate_continuation_error_messages() const
+{
+    auto line_no = m_current_logical_line_source.begin_line;
+    for (const auto& s : m_current_logical_line.segments)
+    {
+        if (s.continuation_error)
+        {
+            parsing::parser_error_listener_ctx listener(*m_ctx->hlasm_ctx, nullptr);
+            listener.add_diagnostic(
+                diagnostic_op::error_E001(range { { line_no, 0 }, { line_no, s.code_offset_utf16 } }));
+            collect_diags_from_child(listener);
```

Comment:
Why not make opencode_provider inherit from diagnosable_ctx? Then these lines would be just:
```suggestion
            add_diagnostic(
                diagnostic_op::error_E001(range { { line_no, 0 }, { line_no, s.code_offset_utf16 } }));
```